src/IOStream.cpp

```diff
@@ -0,0 +1,164 @@
+*/
+/* The original authors have included a copy of the license specified above in the
+   'Docs' folder. See 'gpl.txt'
+*/
+/* We welcome the use of the Mezzanine engine to anyone, including companies who wish to
+   Build professional software and charge for their product.
+
+   However there are some practical restrictions, so if your project involves
+   any of the following you should contact us and we will try to work something
+   out:
+    - DRM or Copy Protection of any kind(except Copyrights)
+    - Software Patents You Do Not Wish to Freely License
+    - Any Kind of Linking to Non-GPL licensed Works
+    - Are Currently In Violation of Another Copyright Holder's GPL License
+    - If You want to change our code and not add a few hundred MB of stuff to
+        your distribution
+
+   These and other limitations could cause serious legal problems if you ignore
+   them, so it is best to simply contact us or the Free Software Foundation, if
+   you have any questions.
+
+   Joseph Toppi - toppij@gmail.com
+   John Blackwood - makoenergy02@gmail.com
+*/
+
+#include "IOStream.h"
+
+#include <cstring>
+
+namespace {
+    /// @brief An enum to store frequently used constants for stream operations.
+    enum Stream_Constant
+    {
+        Temp_Size = 128
+    };
+}
+
+namespace Mezzanine
+{
+    IOStream::IOStream(std::streambuf* Buf) :
+        std::iostream(Buf)
+        {  }
+
+    ///////////////////////////////////////////////////////////////////////////////
+    // Stream Base Operations
+
+    Boole IOStream::EoF() const
+        { return this->eof(); }
+
+    Boole IOStream::Bad() const
+        { return this->bad(); }
+
+    Boole IOStream::Fail() const
+        { return this->fail(); }
+
+    Boole IOStream::IsValid() const
+        { return this->good(); }
+
+    void IOStream::ClearErrors()
+        { this->clear(); }
+
+    ///////////////////////////////////////////////////////////////////////////////
+    // Input methods
+
+    size_t IOStream::Read(void* Buffer, const StreamSize Size)
+    {
+        this->read(static_cast<char*>(Buffer),Size);
+        return static_cast<size_t>( this->gcount() );
+    }
+
+    size_t IOStream::ReadLine(Char8* Buffer, const StreamSize Size, const Char8 Delim)
+    {
+        this->getline(Buffer,Size,Delim);
+        return static_cast<size_t>( this->gcount() );
+    }
+
+    void IOStream::SetReadPosition(StreamPos Position)
+        { this->seekg(Position); }
+
+    void IOStream::SetReadPosition(StreamOff Offset, SeekOrigin Origin)
+        { this->seekg(Offset,static_cast<std::ios_base::seekdir>(Origin)); }
+
+    StreamPos IOStream::GetReadPosition()
+        { return this->tellg(); }
+
+    Boole IOStream::Sync()
+        { return ( this->sync() == 0 ); }
+
+    String IOStream::GetAsString()
+    {
+        String Ret;
+        size_t RetSize = static_cast<size_t>( this->GetSize() );
+        if( RetSize > 0 ) {
+            Ret.reserve(RetSize);
```

Comment:
+1 for the null terminator?

Reply:
https://stackoverflow.com/questions/30111288/stdstringreserve-and-end-of-string-0